### ISSUE TITLE
Missing tests for deterministic signatures

### ISSUE THREAD

**Issue by afk11** [phpecc/phpecc#61]
When I saw https://github.com/mdanter/phpecc/issues/47 I remembered that there are test vectors in RFC6979 for some of the curves we have here. It didn't include secp256k1, but I found some that two implementations were in agreement on. 

I'll look for these, want to mark this for 0.3.0?

**Comment by aztech-dev** [comment]
Hooray for more tests :)

**Comment by aztech-dev** [comment]
Wouldnt https://tools.ietf.org/html/rfc6979#appendix-A.2 be a good startring point ?

**Comment by aztech-dev** [comment]
@afk11 can you have a look at https://github.com/aztech-forks/php-ecc/commit/1359e05a413bff024f62ef0dba70a7d0ec147577 ?

Can't get the tests to pass using data from https://tools.ietf.org/html/rfc6979#appendix-A.2.3

Am I missing something ?

**Comment by aztech-dev** [comment]
More specifically, the testsuite passes with the vectors you have in `rfc6979.json` (I just ported them to the YAML files), but not with the vectors from RFC6979

**Comment by afk11** [comment]
Yeah you're right :/ I tried them and got the same. I was hoping to get to this earlier but things got busy. On it now!

**Comment by afk11** [comment]
For reference my test vectors were taken from here, as a few implementations had used them and agreed with eachother: https://bitcointalk.org/index.php?topic=285142.msg3150733#msg3150733

**Comment by aztech-dev** [comment]
Seems like they're also used here: https://github.com/warner/python-ecdsa/blob/master/ecdsa/test_pyecdsa.py

Golang also some reference vectors https://golang.org/src/crypto/ecdsa/testdata/

I'll try to port those vectors into the test suite tonight (and cover all curves, I currently only did NIST-P192). Maybe that'll help tracking down the issue.

**Comment by aztech-dev** [comment]
Also, considering the tests fail when checking k, the issue is most likely in the HMAC rng, though I haven't had time to investigate.. Considering it's your code and you already know the RFC, I'll leave you to investigate :p

**Comment by aztech-dev** [comment]
On an interesting side note, I re-ran the signature tests, but using k value from the test vectors instead of the generated one -- there are still two failing tests:

```
There were 2 failures:

1) Mdanter\Ecc\Tests\Curves\SpecBasedCurveTest::testHmacSignatures with data set #5 ('nist-p192', Mdanter\Ecc\GeneratorPoint Object (...), '6FAB034934E4C0FC9AE67F5B5659A...E09FD4', 'sha224', 'sample', '4381526B3FC1E7128F202E1945055...F015D8', 'A1F00DAD97AEEC91C95585F36200C...0378F5', 'E07EC1304C7C6C9DEBBE980B96926...A0F97A')
s
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'5504604701309298433102410446644795707377107352225209973114'
+'2891875328107479582659734658798991073926596214052831394172'

/home/thibaud/Documents/Workspaces/forks/php-ecc/tests/unit/Curves/SpecBasedCurveTest.php:165
phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:151
phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:103

2) Mdanter\Ecc\Tests\Curves\SpecBasedCurveTest::testHmacSignatures with data set #6 ('nist-p192', Mdanter\Ecc\GeneratorPoint Object (...), '6FAB034934E4C0FC9AE67F5B5659A...E09FD4', 'sha256', 'sample', '32B1B6D7D42A05CB449065727A848...261496', '4B0B8CE98A92866A2820E20AA6B75...5ECB55', 'CCDB006926EA9565CBADC840829D8...381B85')
s
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'5023041631781708045212851554060961543112660311254607862661'
+'1236024623523781542929074994282459219226838443571985200861'

```

**Comment by afk11** [comment]
Yeah it probably is the HMAC. I think the issue was I wasn't expanding/shrinking some of the strings as the RFC prescribes.. Still working on something for this. I tried the vectors for nist-p256/sha256 and nist-p224/sha224 and they worked, seems to only fail when the size differs now!

**Comment by afk11** [comment]
Funny those are the s parameters.. I am looking at the same thing here, how can s fail if we have the right r? 
Left is output, right is the expected result
```
Try nist-p192 / sha1 / 'sample' 
k: 37d7ca00d2c7b0e5e412ac03bd44ba837fdd5b28cd3b0021 (37D7CA00D2C7B0E5E412AC03BD44BA837FDD5B28CD3B0021)
R: 98c6bd12b23eaf5e2a2045132086be3eb8ebd62abf6698ff (98C6BD12B23EAF5E2A2045132086BE3EB8EBD62ABF6698FF)
S: 57a22b07dea9530f8de9471b1dc6624472e8e2844bc25b64 (57A22B07DEA9530F8DE9471B1DC6624472E8E2844BC25B64)
Try nist-p192 / sha224 / 'sample'
k: 4381526b3fc1e7128f202e194505592f01d5ff4c5af015d8 (4381526B3FC1E7128F202E194505592F01D5FF4C5AF015D8)
R: a1f00dad97aeec91c95585f36200c65f3c01812aa60378f5 (A1F00DAD97AEEC91C95585F36200C65F3C01812AA60378F5)
S: 75f09652acfacdf1a81835e61f81e1c9623ec67b9374797c (E07EC1304C7C6C9DEBBE980B9692668F81D4DE7922A0F97A)

```

**Comment by aztech-dev** [comment]
I don't know actually... Something rather odd is going on, but I haven't been able to figure out what yet. I compared the signing code with the one in python's ECDSA implementation, and it looks allright (though I haven't investigated thoroughly).

The latest version I have here https://github.com/aztech-forks/php-ecc/compare/test-vectors fails with R's also... I need to clear up some time to look into it.

**Comment by aztech-dev** [comment]
So far I have the following combos fail:

- NIST-P192 / sha224
- NIST-P192 / sha256
- NIST-P224 / sha256

With the test vectors from above branch. I'm suspecting something like string padding (or an extra or missing 0 in hex) being the root issue, considering the algo isn't failing for all curve/hash combinations.

**Comment by afk11** [comment]
Padding is definitely at root here for some of the failures, I've added some stuff locally to make it comply more strictly with the RFC with how it deals with string sizes.. Think I have one last thing to address on that, though will push shortly for you to have a look.

The RFC says (For the int2octets) if your Hmac output is too short, or too large, it must be padded with '0' *bits!*, or truncated to the right size. At the moment this is tricky, since base conversions drop any leading zeroes (and having gone from base 10 -> base 2 -> base 16 and then pack.... 

I'm also planning to stop testing r/s until all the k vectors pass, to avoid being upset by bizarre results ;) I'm not sure what to say about S being wrong. That would mean it's the private key or message hash, but, both of these are used to generate K which is correct...

**Comment by aztech-dev** [comment]
If you get k to work, but r or s still failing (they obvisouly will, since they fail when using k from test data set instead of the calculated one), there are not that many options left considering the operations involved in sign:

- The sign algo itself is wrong (though it did not look like it)
- The mul algo is failing at some point (though I would expect at least one private/public key test to fail in that case -- I also did a test run with the mul algo reverted to the original implementation, and the tests were failing all the same)
- The GMP lib and/or adapter is failing at some point
- The test data itself is incorrect (highly doubt it)
- The test case has a bug (possible)

**Comment by aztech-dev** [comment]
Ok, found the issue. Turns out hash wasn't truncated before signing (step 2 [here](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm)), which was why the tests are failing (noticed that all failing tests were tests were the key size was smaller than the hash size).

I pushed https://github.com/aztech-forks/php-ecc/commit/2d41eddfc869c24a7ce0f66ccfa84d35d8811400 which should be all green on tests.

That does however raise the issue of do we leave the responsibility of properly truncating the hash to the consumer of the library, or do we enforce it in `Signer` class (will require some extra data on curves, so that they carry their keysize) ?

**Comment by aztech-dev** [comment]
cc @btcdrak @rubensayshi what's your opinion on this ? (ref last paragraph of previous comment)

**Comment by afk11** [comment]
True.. The only options are worrying, as sign is a simple enough function. Sign() fails intermittently it seems. Finding R is $G->mul($k)->getX(), which is straightforward and hasn't failed yet. The inverseMod stuff is still simple and entirely handled by GMP but seems to be the part that fails. God help us if that's it :)

You'd really hope the test vectors are correct, but lets assume they're fine for now.. 

There is one other thing I might try later - so because of the symmetry about elliptic curves, given a sig (r,s) by k, for message m, anyone can produce an equally valid signature for this message by doing p - s. 

It's a useful/annoying property if

**Comment by afk11** [comment]
OH! Yes that would do it, christ! That is an interesting one, so the Hmac class was correctly truncating, producing the correct k, but the value being passed to sign was incorrect. 

That makes way more sense. Good job!

**Comment by aztech-dev** [comment]
Yup, and r was correct because it's not dependent upon the hash, whereas s is. Glad it's solved actually... I'll raise another issue or P/R for how to handle hash size in signing.

**Comment by afk11** [comment]
Cool nice one. Think you could give my newly modified Hmac class a try against the test vectors? It takes less shortcuts with padding, and implements the helper functions specified in the document, which should make it easier to review in future.

**Comment by aztech-dev** [comment]
Sure, just a sec. I left a few comments on your P/R, nothing too fancy though.

**Comment by aztech-dev** [comment]
Still get two failing tests on 's' (p192 with sha384 and sha512), but k is good.

I refactored a bit your commits, especially the tests -- you should avoid loops and use the dataProvider annotation, makes it much easier to track failing tests. Changes here https://github.com/aztech-forks/php-ecc/commit/cabb7cb974415a7cd43fd5b9ec106014e9e6dd6c

**Comment by afk11** [comment]
Ah nice! Yeah the tests failing without any indication of which was annoying.. cheers!

I had some helpers to print something if it hit any code that was meant to truncate/expand in int2octets, I think this part hasn't been tested so much. 

I gotta work on some other stuff today - but catch up later on!

**Comment by btcdrak** [comment]
@aztech-dev I suggest getting advice on the `Signer` question in the #bitcoin-dev IRC channel...

**Comment by afk11** [comment]
Since the ECDSA spec calls for it, I would add it to phpecc's signer. It won't execute that code normally I guess, but its important that its there I think.

I think key size is the same as qlen (bit length of the order), will try confirm that in a bit

**Comment by aztech-dev** [comment]
Pushed another commit to my test-vectors branch, it passes all tests now. Failing tests were due to some missing padding on the hash bits.

Now that this issue is cleared out, gonna have to find a clean strategy to enforce proper signatures without some mambo jumbo string manipulation prior to signing...

**Comment by afk11** [comment]
Nice one! Have you added all the other tests? I will axe my implementation in bitcoin now and we can keep here :)

**Comment by afk11** [comment]
I'll probably leave you to finish this as you've refactored the tests/data fixtures already and fixed the bug? :P

**Comment by aztech-dev** [comment]
Yeah, hopefully can handle over the weekend... was thinking of extracting a "Hash" class that takes cares of hashing messages and providing padded strings. 

Also thinking we should "intern" the responsibility of hashing inside both the HMAC rng and the signer class so client code doesn't have that responsibility at all, and would instead only have to pass clear text messages to both classes.

**Comment by afk11** [comment]
Most of the curves have tests for Rfc6979 now, with the exception of secp256r1 and secp385r1..

**Comment by afk11** [comment]
We now have tests for every curve besides nist-p521. I'll track this down sometime soon, so will leave this open.

**Comment by btcdrak** [comment]
bump

**Comment by afk11** [comment]
Fixed in master, and #158 backports the fix to 0.3
